container/go/pkg/compat/reader.go

```diff
@@ -16,69 +16,314 @@
-func ReadWithBaseTarball(tarballPath string, layersPath []string) (v1.Image, error) {
-	img, err := tarball.ImageFromPath(tarballPath, nil)
+// reader maintains the state necessary to build a legacyImage object from an
+// ImageParts object.
+type reader struct {
+	// parts is the ImageParts being loaded.
+	parts ImageParts
+	// baseManifest is the manifest of the very first base image in the chain
+	// of images being loaded.
+	baseManifest *v1.Manifest
+	// config is the config of the image being loaded.
+	config *v1.ConfigFile
+	// layerLookup is a map from the diffID of a layer to the layer
+	// itself.
+	layerLookup map[v1.Hash]v1.Layer
+}
+
+// loadMetadata loads the image metadata for the image parts in the given
+// reader.
+func (r *reader) loadMetadata() error {
+	cf, err := os.Open(r.parts.Config)
+	if err != nil {
+		return errors.Wrapf(err, "unable to open image config file %s", r.parts.Config)
+	}
+	c, err := v1.ParseConfigFile(cf)
+	if err != nil {
+		return errors.Wrapf(err, "unable to parse image config from %s", r.parts.Config)
+	}
+	r.config = c
+	if r.parts.BaseManifest == "" {
+		// Base manifest is optional. It's only needed for images whose base
+		// manifests have foreign layers.
+		return nil
+	}
+	mf, err := os.Open(r.parts.BaseManifest)
 	if err != nil {
-		return nil, errors.Wrapf(err, "unable to parse image from tarball at %s", tarballPath)
+		return errors.Wrapf(err, "unable to open base image manifest file %s", r.parts.BaseManifest)
+	}
+	m, err := v1.ParseManifest(mf)
+	if err != nil {
+		return errors.Wrapf(err, "unable to parse base image manifest from %s", r.parts.BaseManifest)
+	}
+	r.baseManifest = m
+	return nil
+}
+
+// foreignLayer represents a foreign layer usually present in windows images.
+type foreignLayer struct {
+	// digest is the digest of this foreign layer.
+	digest v1.Hash
+	// diffID is the diffID of this foreign layer.
+	diffID v1.Hash
+	// size is the size of the foreign layer.
+	size int64
+	// urls are the urls where the actual contents of the foreign layer can
+	// be downloaded from.
+	urls []string
+}
+
+var _ v1.Layer = (*foreignLayer)(nil)
+
+// DiffID returns the diffID of this foreign layer.
+func (l *foreignLayer) DiffID() (v1.Hash, error) {
+	return l.diffID, nil
+}
+
+// Digest returns the digest of this foreign layer.
+func (l *foreignLayer) Digest() (v1.Hash, error) {
+	return l.digest, nil
+}
+
+// Uncompressed returns a blank reader for this foreign layer.
+func (l *foreignLayer) Uncompressed() (io.ReadCloser, error) {
+	r := bytes.NewReader([]byte{})
+	return ioutil.NopCloser(r), nil
+}
+
+// Compressed returns a blank reader for this foreign layer.
+func (l *foreignLayer) Compressed() (io.ReadCloser, error) {
+	return l.Uncompressed()
+}
+
+// Size returns the number of bytes in the compressed version of this layer.
+func (l *foreignLayer) Size() (int64, error) {
+	return 0, nil
```

Comment:
does not seem right to return here 0 unconditionally?

Reply:
Yes this is intentional. This returns the number of bytes in the blob returned by the `Compressed` function on this layer. Since the `Compressed` function returns an empty blob, the size needs to match or the image tarball will fail to build. The correct size and the location to download the image from is added to the generated manifest in the `genManifest` function in `compat/image.go` file. Docker's image loader is aware of this discrepancy and uses the manifest to process foreign layers instead of the actual layer blob included in the image.

Reply:
The `container_test` target `//tests/container:basic_windows_image_test_go_join_layers` tests this case.

Reply:
If that is the case please update the docs to explain a bit more why this is returning a fixed value.